Allow for building of both shared and static libraries

timkpaine opened the pull request:
Allows for optional building of shared libraries instead of static. Non-vendored test path switched to also build shared libraries, runs tests: https://github.com/timkpaine/UHDM/actions/runs/4351690945

hzeller commented:
What will be the shared object version ? Right now while things are still changing quickly there might be the requirement to link exactly a particular version in Surelog. Surelog 1.48 might not be compatible with UHDM 1.47 as typically new data structures are added in both. So at the very least we need an so-name with the current version ... `libuhdm.so.1.48` 

 There are probably different things we can/should do to make this happen
  * Have the shared object version be the same as the current library version. Right now, this is unfortunately just [a tag](https://github.com/chipsalliance/UHDM/tags) so not necessarily easy to exctract at build-time.
  * We should actually have a public API that emits the version, so that Surelog can call that to figure out if it would be compatible (I can add that). Maybe something like `uint32_t getUHDMVersion()`  that returns `UHDM_VERSION` (see below).
 
For the above, we should have a single place where we configure the version. Probably in the cmake file which then generates a header that contains the version number ? Something along the lines of

```c++
#define UHDM_VERSION_MAJOR 1
#define UHDM_VERSION_MINOR 49
#define UHDM_VERSION ((UHDM_VERSION_MAJOR << 16) | (UHDM_VERSION_MINOR))
```

With `getUHDMVersion()` and the header-defined `UHDM_VERSION`, the caller can be sure that they use the version they want, and that they are compatible with the headers they see.

That is the same version that the API returns on `getVersion()` of sorts as well as the so name and the tag/release on github. Something similar we need in Surelog. @timkpaine , since you have good packaging experience, is there an approach you'd recommend ? Maybe something generated from cmake ?

hzeller commented:
Looks like https://cmake.org/cmake/help/latest/command/configure_file.html might be a good candidate.

timkpaine commented:
this is my preferred method, also can export lots of helpful build info via `#cmakedefine`

hzeller commented:
So if we define the version in the CMakeLists file, we can
  * generate a `uhdm-version.h`  with everything we need (a header visible to the user of the library. And also a UHDM library implementation of `getUHDMVersion()` ).
  * Have the version number in the cmake file to create the shared object version number
  * And _maybe_ even could auto-`git tag` it in the CI action (leaving it to the maintainers to make it an actual Release from a Tag but for software packaging folks a tag is already sufficiently valuable most of the time)

I like that; would you like to make that happen @timkpaine ? (Sorry to load that on you, but I think it might be something you can do so that it won't slow down your current pace in getting the targets ready for packaging). I can also work on that, but can't promise much for the next 2-3 days.

timkpaine commented:
ha no problem, i can do it. This all started because i wanted to get https://github.com/antmicro/yosys-systemverilog working on my mac worker machines, that made me generalize a bunch of stuff in https://github.com/chipsalliance/yosys-f4pga-plugins. When i went to make the changes there, I realized its using its own conda stack so i wanted to get yosys and co into conda-forge and homebrew, but both of those orgs are pretty staunchly against vendoring dependencies and static linking, and both require version numbers. So ultimately i'd need these changes anyway, happy to do them in the next few days and we can discuss further.

hzeller commented:
I've already packaged UHDM/Surelog/yosys_f4pga_plugins on nixos (around August), but there I am also missing the de-vendoring, that is why you see some of my TODOs in the CMakeLists.
So looking forward for this to happen so that I can update my packages as well.

Thomasb81 commented:
> So if we define the version in the CMakeLists file, we can
> 
>     * generate a `uhdm-version.h`  with everything we need (a header visible to the user of the library. And also a UHDM library implementation of `getUHDMVersion()` ).

Probably a dummy question, would this allow to check that the set of tool use to create and later the set of tool that will use the .uhdm file use the exact same flatbuffers schema ?
Is that a build-in feature of flatbuffers ? I feel to already face odd behavior because of this kind of miss-match.

hzeller commented:
I think as soon as we have a proper version number in the API, we should always emit it when writing the file - then we can check that it is the same. I agree, this is an important point @Thomasb81 .

Possibly this should even be something internal to the UHDM lib, which will fail on loading a file and reports something like '_an't read xzy.uhdm, need version in range 1.45..1.48 for that_ ... or something along those lines.

timkpaine commented:
open to comments on what's here. One thing I do not like is that it presumes a Git repo. This is convenient for us when developing (e.g. just tag and it will automatically pick up the version), but this means you can't build direct from the tarball (so the release assets won't be useable).

This is probably fine for now since the official installation docs say to clone, but if we have more stable version and automated deployment, the version should exist outside of git.

timkpaine commented:
> Possibly this should even be something internal to the UHDM lib, which will fail on loading a file and reports something like '_an't read xzy.uhdm, need version in range 1.45..1.48 for that_ ... or something along those lines.

I've done this in other contexts with an "ABI" version parallel to the library version (semantics of ABI here being abused to simply mean "some form of library compatibility").

hzeller commented on CMakeLists.txt:
```diff
@@ -174,9 +174,47 @@ foreach(src_file ${uhdm-GENERATED_SRC})
+else()
+  # if set in env, assume some other deployment mechanism where
+  # not working from the git repo / where git not installed
+  set(UHDM_VERSION_MAJOR ENV{UHDM_VERSION_MAJOR})
```
What happens if a user just downloads the source archive and compiles ? It looks like in that case they might need to set the environment variables correctly.

Can we set the reasonable defaults in the CMakeFile directly ? Maybe providing a version directly in the [`project()` cmake](https://cmake.org/cmake/help/latest/command/project.html) function (around line 26)

```cmake
project(UHDM VERSION 1.48)
```

Now if `git` is available, we still do the extracting from the git repo with the exact git tag and sha hash.
(we could generate a warning relevant for developers if the tag returns a different major/minor than we expect in the project version).

If git does not exist, we configure the `UHDM_VERSION_MAJOR`, `UHDM_VERSION_MINOR` which is automatically set by the `project()` call. and set the `UHDM_VERSION_COMMIT_SHA` to `"unknown"` or something similar.

hzeller replied:
Yes, cmake version should be source of truth.
The current tagging process (which still is manually) also uses the source to deduce the version number (but only in Surelog currently).

Goal is to have the tag happening automatically at CI time, never manually.

timkpaine replied:
sounds good

hzeller reviewed (commented):
Very nice. I think we can make the version a bit easier for folks downloading the source repo by adding the version directly into there. Makes things a bit more self-contained.

hzeller commented on CMakeLists.txt:
```diff
@@ -329,3 +364,19 @@ install(DIRECTORY ${GENDIR}/uhdm/
+include(CMakePackageConfigHelpers)
+
+# generate the config file that is includes the exports
+configure_package_config_file(
```
Maybe not in this PR, but in a follow-up: Can we also generate a [pkg-config](https://www.freedesktop.org/wiki/Software/pkg-config/) compatible [file](https://manpages.ubuntu.com/manpages/jammy/man5/pc.5.html) ? It is generally a format used in projects that do not use cmake.

 For instance [yosys-f4pga-plugins](https://github.com/chipsalliance/yosys-f4pga-plugins/blob/main/systemverilog-plugin/Makefile) uses plain makefiles, determining the headers and defines with `CXXFLAGS+=$(pkg-config --cflags uhdm)` and `LDFLAGS+=$(pkg-config --libs uhdm)` there would probably be very nice and improve packageability.

timkpaine replied:
looking into this now

hzeller commented on Config.cmake.in:
```diff
@@ -0,0 +1,3 @@
+@PACKAGE_INIT@
+
+include ( "${CMAKE_CURRENT_LIST_DIR}/UHDM.cmake" )
```
does the UHDM.cmake file be part of the PR, or will it be somehow generated ?

timkpaine commented:
Updated to pull version information from `project`. If building out of the git repo (e.g. in CI), i added a check in the CMakeList to ensure that last git tag == project version, and tested with a typo:

```
CMake Error at CMakeLists.txt:200 (message):
  Git tag 1.{UHDM_VERSION_MINOR} does not match project version 1.48.  Did
  you forget to update project version in CMakeLists.txt?
```

timkpaine commented:
looks good
```
> pkg-config UHDM --cflags
-I/usr/local/include/uhdm

> pkg-config UHDM --libs
-L/usr/local/lib/uhdm -luhdm

```

Thomasb81 commented on CMakeLists.txt:
```diff
@@ -174,9 +176,52 @@ foreach(src_file ${uhdm-GENERATED_SRC})
 add_library(uhdm ${uhdm-GENERATED_SRC})
 set_target_properties(uhdm PROPERTIES PUBLIC_HEADER ${GENDIR}/uhdm/uhdm.h)
 
+if(BUILD_SHARED_LIBS)
```
Does this allow to properly compile uhdm as shared library with in particular `capnproto ` to be compile with `-fPIC` switch ?

At my end I try to work on a uhdm python wrapper https://github.com/Thomasb81/UHDM/tree/python_wrapper
But if I mimic correctly this in my working environment it failed to properly compile (I may be wrong at doing this):
```
[100%] Built target uhdm-hier
[100%] Linking CXX shared module ../bin/_swig_example.so
/usr/bin/ld: ../third_party/capnproto/c++/src/kj/libkj.a(exception.c++.o): relocation R_X86_64_TPOFF32 against `_ZN2kj12_GLOBAL__N_1L19threadLocalCallbackE' can not be used when making a shared object; recompilé avec -fPIC
```

I see in antlr project that instead to control static/shared library they define 2 targets one for static and the second for shared. This allow end application to choose.
The drawback is project build without specific target will compile twice the library.

Is there arguments to continue to link uhdm against static library ? It seems to me that linking against shared library could be more efficient regarding elaboration time.

timkpaine replied:
Yes, this PR lets you compile UHDM as a shared library via the cmake flag `BUILD_SHARED_LIBS`. Capnproto and the other dependencies have their own separate build processes. However, you can see how now when I build `capnproto` in the `main.yml`, I use the cmake flag `CMAKE_POSITION_INDEPENDENT_CODE` which will use the compiler-specific flag (e.g. `-fPIC`). If you want to do a build with vendored sources (e.g. from the `git` repo) you would run a command like: `cmake -B build -DBUILD_SHARED_LIBS=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_POSITION_INDEPENDENT_CODE=ON . && cmake --build build`.

hzeller commented on CMakeLists.txt:
```diff
@@ -174,9 +176,52 @@ foreach(src_file ${uhdm-GENERATED_SRC})
 set_target_properties(uhdm PROPERTIES PUBLIC_HEADER ${GENDIR}/uhdm/uhdm.h)
 
+if(BUILD_SHARED_LIBS)
+  set_property(TARGET uhdm PROPERTY POSITION_INDEPENDENT_CODE 1)
```
Is there a way to set the soname version, so that we end up with libuhdm.so.1.48 as well as symbolic links libuhdm.so.1 and libuhdm.so in an install target ?

hzeller reviewed (approved):
I think this looks very good and ready to commit. If there are little things here and there to improve later, that can go int later pull rquests.

I have one little question if/how we can automatically make sure to have soname and version ready, but that can be done separately.

@alaindargelas , handing to you

alaindargelas commented:
@hzeller @timkpaine ,
I merged this PR, then checkout out in Surelog.
What am I supposed to do here to get this to compile?



~/Surelog$ make
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr/local -DBUILD_SHARED_LIBS=OFF  -S . -B build
-- Proceeding with version: 23.3.3.10
-- CMAKE_CXX_FLAGS: 
-- Building without demo. To enable demo build use: -DWITH_DEMO=True
-- Found Python3: /usr/bin/python3.9 (found version "3.9.7") found components: Interpreter 
-- Python3_EXECUTABLE = /usr/bin/python3.9
CMake Error at third_party/UHDM/CMakeLists.txt:205 (message):
  Git tag 1.48 does not match project version ..  Did you forget to update
  project version in CMakeLists.txt?


-- Configuring incomplete, errors occurred!
make: *** [Makefile:46: run-cmake-release] Error 1

timkpaine commented:
> * Existence of a ".git" folder doesn't mean that git command is available from within the CMake file/process.

It's a fairly close assumption, if you have a .git folder you've likely cloned (vs pulling the release tarball), and easier to rely on `git` than using a find to try to grab git cmake variables. 

> * Surelog is not always the root project. This is introducing a restraint for user/client applications to provide a version

It shouldn't, in fact the main goal here is so that you can build and deploy UHDM on its own to e.g. homebrew, conda-forge, etc.


> * What happens if a commit is introduced in a submodule, say UHDM, and another one in Surelog. How are the versions going to match in this case?

However one would normally do it, Surelog should do a FindUHDM(<version>) when not vendoring. When vendoring, we should use the commit corresponding to the last tagged release.

> Is there really a need to run all the versioning logic for local/dev builds?

Only if you're ok with being careful during the release, again you don't want the git tag to be 1.50 while the cmakelists says 1.49

timkpaine commented:
i have no problem for that, just means the CI will have to be responsible for version mgmt and tagging

hzeller commented:
Yes, the CI should be responsible for the tagging. It should take the version number from the cmake version and create a new tag whenever that changes. So as soon as someone checks in a new version with an updated cmake project version that will automatically in the CI be used to tag that version in git.

I'll add that to the CI in UHDM (since we have the project version in there now), and once we have the version number also in Surelog CMake project, there as well (I'll first play with a toy project first).

For now, I think we should always take the version number from the CMake project and leave out the test for the tag; so if git is available, we could stil luse that to augment the hash.
